dask/dataframe/io/parquet/core.py

```diff
@@ -869,11 +869,13 @@ def to_parquet(
                 ):
                     path_with_slash = path.rstrip("/") + "/"  # ensure trailing slash
                     for input in layer.inputs:
-                        if input["piece"][0].startswith(path_with_slash):
-                            raise ValueError(
-                                "Reading and writing to the same parquet file within the "
-                                "same task graph is not supported."
-                            )
+                        # Note that `input` may be either `dict` or `List[dict]`
+                        for piece_dict in input if isinstance(input, list) else [input]:
```

Comment:
It feels like this is very fragile. IO is always messy, I agree, but should the construction of the layers have normalised things such that consumers of the layers don't need to dispatch on isinstance? Another way of asking this question is whether any other inspectors of the layer object need to care about this `dict | list[dict]` distinction?

Reply:
Yes, this is indeed messy, and it would be nice if we could **require**  this to always be `List[dict]` to make it clear that a `read_partition` call will be expected to read from 1+ files. These two possibilities are certainly handled in the read code, but it'd tough to be sure about *other* code that is attempting to inspect the data structures that are created within `read_parquet`.